db/cache_db.go

```diff
@@ -0,0 +1,227 @@
+package db
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"sync/atomic"
+)
+
+// If value is nil but deleted is false,
+// it means the parent doesn't have the key.
+// (No need to delete upon Write())
+type cDBValue struct {
+	value   []byte
+	deleted bool
+}
+
+// CacheDB wraps an in-memory cache around an underlying DB.
+type CacheDB struct {
```

Comment:
We really should have two caches, a read cache and a write cache... Otherwise, if we query a value, we write that to the db again. Not only is this wasteful use of IO, but with the version tree, this actually changes the hash.

https://github.com/cosmos/cosmos-sdk/blob/develop/state/kvcache.go#L6-L11

Issue:
https://github.com/cosmos/gaia/issues/41
https://github.com/cosmos/cosmos-sdk/pull/271

Reply:
Fixing so that we have a "dirty" flag.